driver/src/main/com/mongodb/GridFSDownloadStreamImpl.java

```diff
@@ -0,0 +1,171 @@
+    private final int chunkSizeInBytes;
+    private final int numberOfChunks;
+    private int chunkIndex;
+    private int bufferOffset;
+    private long currentPosition;
+    private byte[] buffer = null;
+    private boolean closed;
+
+    GridFSDownloadStreamImpl(final BsonDocument fileInfo, final MongoCollection<BsonDocument> chunksCollection) {
+        this.fileInfo = notNull("file information", fileInfo);
+        this.chunksCollection = notNull("chunks collection", chunksCollection);
+
+        try {
+            isTrue("file information contains _id", fileInfo.containsKey("_id") && fileInfo.get("_id").isObjectId());
+            isTrue("file information contains length", fileInfo.containsKey("length") && fileInfo.get("length").isInt64());
+            isTrue("file information contains chunkSize", fileInfo.containsKey("chunkSize") && fileInfo.get("chunkSize").isInt32());
+        } catch (IllegalStateException e) {
+            throw new MongoGridFSException("GridFS file information is not in the expected format", e);
+        }
+        fileId = fileInfo.getObjectId("_id");
+        lengthInBytes = fileInfo.getInt64("length").getValue();
+        chunkSizeInBytes = fileInfo.getInt32("chunkSize").getValue();
+        numberOfChunks = (int) Math.ceil((double) lengthInBytes / chunkSizeInBytes);
+    }
+
+    @Override
+    public BsonDocument getFileInformation() {
+        return fileInfo;
+    }
+
+    @Override
+    public InputStream getInputStream() {
+        return this;
+    }
+
+    @Override
+    public int read() throws IOException {
+        byte[] b = new byte[1];
+        int res = read(b);
+        if (res < 0) {
+            return -1;
+        }
+        return b[0] & 0xFF;
+    }
+
+    @Override
+    public int read(final byte[] b, final int off, final int len) throws IOException {
+        checkClosed();
+
+        if (currentPosition == lengthInBytes) {
+            return -1;
+        } else if (buffer == null) {
+            buffer = getBuffer(chunkIndex);
+            bufferOffset = 0;
+        } else if (bufferOffset >= buffer.length) {
+            chunkIndex += 1;
+            buffer = getBuffer(chunkIndex);
+            bufferOffset = 0;
+        }
+
+        int r = Math.min(len, buffer.length - bufferOffset);
+        System.arraycopy(buffer, bufferOffset, b, off, r);
+        bufferOffset += r;
+        currentPosition += r;
+        return r;
+    }
+
+    @Override
+    public long skip(final long bytesToSkip) throws IOException {
+        checkClosed();
+        if (bytesToSkip <= 0) {
+            return 0;
+        }
+
+        long skippedPosition = currentPosition + bytesToSkip;
+        bufferOffset = (int) skippedPosition % chunkSizeInBytes;
+        if (skippedPosition > lengthInBytes) {
+            chunkIndex = numberOfChunks - 1;
+            currentPosition = lengthInBytes;
+            buffer = null;
+            return skippedPosition - (skippedPosition - lengthInBytes);
```

Comment:
This looks wrong, as it doesn't take into account the previous current position.  Say you have a file that is length 3 and current position = 0, then according to the skip contract:

   expect:
   stream.skip(100) == 3
   steam.skip(100) == 0